Release branch for v0.32.1 (#3787)
tendermint/tendermint: CHANGELOG.md

# Changelog

## v0.32.1

*July 12, 2019*

Special thanks to external contributors on this release: 
@ParthDesai, @climber73, @jim380, @ashleyvega

This release contains breaking changes to our libs folder and ABCI namely:

> - Add a "Recheck Tx" indicator 

Reviewer: ```suggestion
- CheckTx requests include a `CheckTxType` enum that can be set to `Recheck` to indicate to the application that this transaction was already checked/validated and certain expensive operations (like checking signatures) can be skipped
```

> - Removal of `db/debugDB`, `common/colors.go`, `errors/errors.go`, `common/heap.go`, `date.go`, `io.go` and

Reviewer 2: I would remove this and expand on RecheckTx indicator instead (how applications can use it with examples)

> - Removal of `db/debugDB`, `common/colors.go`, `errors/errors.go`, `common/heap.go`, `date.go`, `io.go` and

Reviewer: `Removed various 
```suggestion
- Removed various functions from `libs` pkgs
```

  `GoPath()`, `Prompt()` and `IsDirEmpty()` functions from `os.go`

Friendly reminder, we have a [bug bounty
program](https://hackerone.com/tendermint).

### BREAKING CHANGES:

> - CLI/RPC/Config

Reviewer 2: no need for empty sections here

- Apps

- Go API

>   - [abci] [\#2127](https://github.com/tendermint/tendermint/issues/2127) ABCI / mempool: Add a "Recheck Tx" indicator. Breaks the ABCI

Reviewer: The breaking change here should be described more clearly as introducing the RequestCheckTx and RequestDeliverTx types as the args for those methods in the `abci.Application` interface. Ie. this is specifically about the breaking change to the Go API.

Adding the CheckTxType field is something that should be logged as an `[abci]` feature. It also corresponds to a bump in the ABCI version (see comment below).

    client interface (`abcicli.Client`) to allow for supplying the ABCI
    `types.RequestCheckTx` and `types.RequestDeliverTx` structs, and lets the
    mempool indicate to the ABCI app whether a CheckTx request is a recheck or
    not.
  - [libs] Remove unused `db/debugDB` and `common/colors.go` & `errors/errors.go` files (@marbar3778)
  - [libs] [\#2432](https://github.com/tendermint/tendermint/issues/2432) Remove unused `common/heap.go` file (@marbar3778)
  - [libs] Remove unused `date.go`, `io.go`. Remove `GoPath()`, `Prompt()` and `IsDirEmpty()` functions from `os.go` (@marbar3778)

- Blockchain Protocol

- P2P Protocol

### FEATURES:

> - [node] Refactor `NewNode` to use functional options to make it more flexible

Reviewer: ```suggestion
- [node] Add variadic argument to `NewNode` to support functional options, allowing the Node to be more easily customized. 
```

>   and extensible in the future.

Reviewer: ```suggestion
```

- [node][\#3730](https://github.com/tendermint/tendermint/pull/3730) Add `CustomReactors` option to `NewNode` allowing caller to pass
  custom reactors to run inside Tendermint node (@ParthDesai)

### IMPROVEMENTS:

- [rpc] [\#3700](https://github.com/tendermint/tendermint/issues/3700) Make possible to set absolute paths for TLS cert and key (@climber73)

### BUG FIXES:

- [p2p] [\#3338](https://github.com/tendermint/tendermint/issues/3338) Prevent "sent next PEX request too soon" errors by not calling
  ensurePeers outside of ensurePeersRoutine

> - [behaviour] Return correct reason in MessageOutOfOrder (@jim380)

Reviewer: Missing an issue/PR. 

Also why isn't the `behaviour` pkg under `p2p` cc @brapse
